// @sonoransoftware/sonoran.js/dist/structures/CADActiveUnit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CADActiveUnit = void 0;
const Base_1 = require("./Base");
class CADActiveUnit extends Base_1.Base {
    constructor(instance, data) {
        super(instance);
        this.id = -1;
        this.accId = '';
        this.status = 0;
        this.isPanic = false;
        this.location = '';
        this.aop = '';
        this.isDispatch = false;
        this.data = {
            apiId1: '',
            apiId2: '',
            unitNum: '',
            name: '',
            district: '',
            department: '',
            subdivision: '',
            rank: '',
            group: ''
        };
        if (data)
            this._patch(data);
    }
    _patch(data) {
        if (data.id !== undefined)
            this.id = data.id;
        if (data.accId !== undefined)
            this.accId = data.accId;
        if (data.status !== undefined)
            this.status = data.status;
        if (data.isPanic !== undefined)
            this.isPanic = data.isPanic;
        if (data.location !== undefined)
            this.location = data.location;
        if (data.aop !== undefined)
            this.aop = data.aop;
        if (data.isDispatch !== undefined)
            this.isDispatch = data.isDispatch;
        if (data.data !== undefined) {
            if (data.data.apiId1 !== undefined)
                this.data.apiId1 = data.data.apiId1;
            if (data.data.apiId2 !== undefined)
                this.data.apiId2 = data.data.apiId2;
            if (data.data.unitNum !== undefined)
                this.data.unitNum = data.data.unitNum;
            if (data.data.name !== undefined)
                this.data.name = data.data.name;
            if (data.data.district !== undefined)
                this.data.district = data.data.district;
            if (data.data.department !== undefined)
                this.data.department = data.data.department;
            if (data.data.subdivision !== undefined)
                this.data.subdivision = data.data.subdivision;
            if (data.data.rank !== undefined)
                this.data.rank = data.data.rank;
            if (data.data.group !== undefined)
                this.data.group = data.data.group;
        }
    }
}
exports.CADActiveUnit = CADActiveUnit;
