// @sonoransoftware/sonoran.js/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uuidRegex = exports.CMSSubscriptionVersionEnum = exports.CADSubscriptionVersionEnum = exports.productEnums = void 0;
var productEnums;
(function (productEnums) {
    productEnums[productEnums["CAD"] = 0] = "CAD";
    productEnums[productEnums["CMS"] = 1] = "CMS";
})(productEnums = exports.productEnums || (exports.productEnums = {}));
var CADSubscriptionVersionEnum;
(function (CADSubscriptionVersionEnum) {
    CADSubscriptionVersionEnum[CADSubscriptionVersionEnum["FREE"] = 0] = "FREE";
    CADSubscriptionVersionEnum[CADSubscriptionVersionEnum["STARTER"] = 1] = "STARTER";
    CADSubscriptionVersionEnum[CADSubscriptionVersionEnum["STANDARD"] = 2] = "STANDARD";
    CADSubscriptionVersionEnum[CADSubscriptionVersionEnum["PLUS"] = 3] = "PLUS";
    CADSubscriptionVersionEnum[CADSubscriptionVersionEnum["PRO"] = 4] = "PRO";
    CADSubscriptionVersionEnum[CADSubscriptionVersionEnum["ONE"] = 6] = "ONE";
})(CADSubscriptionVersionEnum = exports.CADSubscriptionVersionEnum || (exports.CADSubscriptionVersionEnum = {}));
var CMSSubscriptionVersionEnum;
(function (CMSSubscriptionVersionEnum) {
    CMSSubscriptionVersionEnum[CMSSubscriptionVersionEnum["FREE"] = 0] = "FREE";
    CMSSubscriptionVersionEnum[CMSSubscriptionVersionEnum["STARTER"] = 1] = "STARTER";
    CMSSubscriptionVersionEnum[CMSSubscriptionVersionEnum["STANDARD"] = 2] = "STANDARD";
    CMSSubscriptionVersionEnum[CMSSubscriptionVersionEnum["PLUS"] = 3] = "PLUS";
    CMSSubscriptionVersionEnum[CMSSubscriptionVersionEnum["PRO"] = 4] = "PRO";
    CMSSubscriptionVersionEnum[CMSSubscriptionVersionEnum["ONE"] = 6] = "ONE";
})(CMSSubscriptionVersionEnum = exports.CMSSubscriptionVersionEnum || (exports.CMSSubscriptionVersionEnum = {}));
exports.uuidRegex = /[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}/i;
